### PR TITLE
ARROW-15374: [C++][FlightRPC] Add support for MemoryManager in data methods

### PR DESCRIPTION
**Pull request by lidavidm**
This enables using Flight with something like CUDA without having to manually copy data. Given the work around UCX on the mailing list, this would enable alternative backends to optimize based on where data is allocated.

### CODE REVIEW

File: cpp/src/arrow/flight/client.cc [hunk 1]
```diff
@@ -460,6 +462,11 @@ class GrpcIpcMessageReader : public ipc::MessageReader {
       stream_finished_ = true;
       return stream_->Finish(Status::OK());
     }
+
+    if (ARROW_PREDICT_FALSE(!memory_manager_->is_cpu() && data->body)) {
+      ARROW_ASSIGN_OR_RAISE(data->body, Buffer::ViewOrCopy(data->body, memory_manager_));
+    }
```
Comment (cyb70289) [inline cpp/src/arrow/flight/client.cc:468]: Nit: Does it make sense to move this code inside peekable_reader_->Next()?

**Review by cyb70289 — APPROVED** [review]
LGTM

File: cpp/src/arrow/flight/flight_cuda_test.cc [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+namespace flight {
+
+Status CheckBuffersOnDevice(const Array& array, const Device& device) {
+  for (const auto& buffer : array.data()->buffers) {
+    if (!buffer) continue;
+    if (!buffer->device()->Equals(device)) {
+      return Status::Invalid("Expected buffer on device: ", device.ToString(),
+                             ". Was allocated on device: ", buffer->device()->ToString());
+    }
+  }
```
Comment (pitrou) [inline cpp/src/arrow/flight/flight_cuda_test.cc:40]: Should we also recurse into child data here?

Comment (pitrou) [reply cpp/src/arrow/flight/flight_cuda_test.cc:40]: Ok, I realize the test might only use primitive data, in which case it doesn't matter.

File: cpp/src/arrow/flight/flight_cuda_test.cc [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+    for (const auto& column : cuda_batch->columns()) {
+      ASSERT_OK(CheckBuffersOnDevice(*column, *device_));
+    }
+
+    ASSERT_OK(writer->WriteRecordBatch(*cuda_batch));
+
+    FlightStreamChunk chunk;
+    ASSERT_OK(reader->Next(&chunk));
+    for (const auto& column : chunk.data->columns()) {
```
Comment (pitrou) [inline cpp/src/arrow/flight/flight_cuda_test.cc:193]: Should we also check the contents of the returned data here?

File: cpp/src/arrow/flight/serialization_internal.cc [hunk 4]
```diff
@@ -272,7 +282,13 @@ grpc::Status FlightDataSerialize(const FlightPayload& msg, ByteBuffer* out,
...
         if (!buffer) continue;
 
-        slices.push_back(SliceFromBuffer(buffer));
+        grpc::Slice slice;
+        auto status = SliceFromBuffer(buffer).Value(&slice);
+        if (ARROW_PREDICT_FALSE(!status.ok())) {
+          // This will likely lead to abort as gRPC cannot recover from an error here
+          return ToGrpcStatus(status);
+        }
+        slices.push_back(slice);
```
Comment (pitrou) [inline cpp/src/arrow/flight/serialization_internal.cc:291]: (not sure how `grpc::Slice` behaves wrt copies)

```suggestion
        slices.push_back(std::move(slice));
```

File: cpp/src/arrow/flight/flight_cuda_test.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/flight_cuda_test.cc:43]: I added a guard against this.

File: cpp/src/arrow/flight/flight_cuda_test.cc (hunk 3 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/flight_cuda_test.cc:217]: Makes sense, I added a helper to do this.

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**